File: android-app/app/src/main/java/com/eshot/app/data/csv/CsvParser.kt
package com.eshot.app.data.csv

import android.util.Log
import com.eshot.app.data.model.Line
import com.eshot.app.data.model.Stop
import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.nio.charset.Charset

class CsvParser {

    fun parseStops(inputStream: InputStream): List<Stop> {
        val (reader, delimiter) = createReaderAndDetectDelimiter(inputStream)
        val stops = mutableListOf<Stop>()
        
        try {
            val headers = reader.readLine()?.split(delimiter)?.map { it.trim().uppercase() } ?: return emptyList()
            // Map header indices
            val idIndex = headers.indexOfFirst { it.contains("ID") }
            val nameIndex = headers.indexOfFirst { it.contains("ADI") || it.contains("NAME") }
            val latIndex = headers.indexOfFirst { it.contains("ENLEM") || it.contains("LAT") }
            val lonIndex = headers.indexOfFirst { it.contains("BOYLAM") || it.contains("LON") }
            val linesIndex = headers.indexOfFirst { it.contains("HATLAR") }

            if (idIndex == -1 || latIndex == -1 || lonIndex == -1) {
                Log.e("CsvParser", "Missing critical columns in Stops CSV. Headers: $headers")
                return emptyList()
            }

            var line = reader.readLine()
            while (line != null) {
                val trimmedLine = line.trim()
                if (trimmedLine.isEmpty()) {
                    line = reader.readLine()
                    continue
                }
                
                val parts = line.split(delimiter).map { it.trim() }
                // Ensure we have enough parts for the critical indices
                val maxRequiredIndex = maxOf(idIndex, latIndex, lonIndex)
                
                if (parts.size > maxRequiredIndex) {
                   try {
                       val lat = parts.getOrNull(latIndex)?.replace(",", ".")?.toDoubleOrNull()
                       val lon = parts.getOrNull(lonIndex)?.replace(",", ".")?.toDoubleOrNull()
                       
                       if (lat != null && lon != null) {
                           stops.add(Stop(
                               id = parts.getOrNull(idIndex) ?: "Unknown",
                               name = if (nameIndex != -1) parts.getOrNull(nameIndex) ?: "Unknown Stop" else "Unknown Stop",
                               latitude = lat,
                               longitude = lon,
                               lineIds = if (linesIndex != -1) parts.getOrNull(linesIndex) ?: "" else ""
                           ))
                       }
                   } catch (e: Exception) {
                       // Skip malformed row
                   }
                }
                line = reader.readLine()
            }
        } catch (e: Exception) {
            Log.e("CsvParser", "Error parsing stops", e)
        } finally {
            reader.close()
        }
        return stops
    }

    fun parseLines(inputStream: InputStream): List<Line> {
        val (reader, delimiter) = createReaderAndDetectDelimiter(inputStream)
        val lines = mutableListOf<Line>()
        
        try {
            val headers = reader.readLine()?.split(delimiter)?.map { it.trim().uppercase() } ?: return emptyList()
            
            val idIndex = headers.indexOfFirst { it == "HAT_NO" || it.contains("ID") }
            val nameIndex = headers.indexOfFirst { it == "HAT_ADI" || it.contains("NAME") }
            val descIndex = headers.indexOfFirst { it.contains("GUZERGAH") }
            val startIndex = headers.indexOfFirst { it.contains("BASLANGIC") }
            val endIndex = headers.indexOfFirst { it.contains("BITIS") }

            if (idIndex == -1) return emptyList()

            var line = reader.readLine()
            while (line != null) {
                val trimmedLine = line.trim()
                if (trimmedLine.isEmpty()) {
                    line = reader.readLine()
                    continue
                }
                
                val parts = line.split(delimiter).map { it.trim() }
                 if (parts.size > idIndex) {
                    lines.add(Line(
                        id = parts.getOrNull(idIndex) ?: "",
                        name = if (nameIndex != -1) parts.getOrNull(nameIndex) ?: "" else "",
                        description = if (descIndex != -1) parts.getOrNull(descIndex) ?: "" else "",
                        startStop = if (startIndex != -1) parts.getOrNull(startIndex) ?: "" else "",
                        endStop = if (endIndex != -1) parts.getOrNull(endIndex) ?: "" else ""
                    ))
                 }
                line = reader.readLine()
            }
        } catch (e: Exception) {
             Log.e("CsvParser", "Error parsing lines", e)
        } finally {
            reader.close()
        }
        return lines
    }

    private fun createReaderAndDetectDelimiter(inputStream: InputStream): Pair<BufferedReader, Char> {
        val bufferedIn = inputStream.buffered()
        bufferedIn.mark(4096)
        
        // Peek
        val buffer = ByteArray(2048)
        val read = bufferedIn.read(buffer)
        bufferedIn.reset()
        
        // Detect Encoding
        var charset = Charset.forName("UTF-8")
        try {
            val decoder = charset.newDecoder()
            decoder.decode(java.nio.ByteBuffer.wrap(buffer, 0, read))
        } catch (e: Exception) {
             // Fallback to Turkish Windows-1254 if UTF-8 fails
             try {
                 charset = Charset.forName("windows-1254")
             } catch (unsupported: Exception) {
                 // If 1254 is not supported on this device, fallback to ISO-8859-9 or default
                 charset = Charset.defaultCharset()
             }
        }
        
        // Detect Delimiter from preview
        val text = String(buffer, 0, read, charset)
        val firstLine = text.lines().firstOrNull() ?: ""
        
        val semicolons = firstLine.count { it == ';' }
        val commas = firstLine.count { it == ',' }
        val tabs = firstLine.count { it == '\t' }
        
        val delimiter = when {
            semicolons >= commas && semicolons >= tabs -> ';'
            tabs > commas -> '\t'
            else -> ','
        }
        
        return Pair(BufferedReader(InputStreamReader(bufferedIn, charset)), delimiter)
    }
}

File: ios-app/IzmirEshot/Parsers/CsvParser.swift
import Foundation

class CsvParser {
    
    func parseStops(data: Data) -> [Stop] {
        guard let content = decodeData(data) else { return [] }
        
        var lines = content.components(separatedBy: .newlines)
        if lines.isEmpty { return [] }
        
        let headerLine = lines.removeFirst()
        let delimiter = detectDelimiter(headerLine)
        
        let headers = headerLine.components(separatedBy: delimiter).map { $0.trimmingCharacters(in: .whitespacesAndNewlines).uppercased() }
        
        guard let idIndex = headers.firstIndex(where: { $0.contains("ID") }),
              let latIndex = headers.firstIndex(where: { $0.contains("ENLEM") || $0.contains("LAT") }),
              let lonIndex = headers.firstIndex(where: { $0.contains("BOYLAM") || $0.contains("LON") }) else {
            return []
        }
        
        let nameIndex = headers.firstIndex(where: { $0.contains("ADI") || $0.contains("NAME") }) ?? -1
        let linesIndex = headers.firstIndex(where: { $0.contains("HATLAR") }) ?? -1
        
        var stops: [Stop] = []
        
        for line in lines {
            let parts = line.components(separatedBy: delimiter)
            if parts.count >= headers.count { // Lenient
                let latIndex = latIndex < parts.count ? latIndex : -1
                let lonIndex = lonIndex < parts.count ? lonIndex : -1
                
                if latIndex != -1, lonIndex != -1,
                   let lat = parts[latIndex].replacingOccurrences(of: ",", with: ".").toDouble(),
                   let lon = parts[lonIndex].replacingOccurrences(of: ",", with: ".").toDouble() {
                    
                    stops.append(Stop(
                        id: parts[idIndex],
                        name: nameIndex != -1 ? parts[nameIndex] : "Unknown",
                        latitude: lat,
                        longitude: lon,
                        lineIds: linesIndex != -1 ? parts[linesIndex] : ""
                    ))
                }
            }
        }
        return stops
    }
    
    private func decodeData(_ data: Data) -> String? {
        // Try UTF-8
        if let utf8 = String(data: data, encoding: .utf8) {
            return utf8
        }
        // Try Windows-1254 (No native constant, use validation or Latin5/ISO-8859-9 approximation)
        // Swift ISOLatin5 (8859-9) is closest to 1254
        if let latin5 = String(data: data, encoding: .isoLatin1) { // Fallback to compatible western
            return latin5 
        }
        // WindowsCP1254 = 0x800004e6 ? No, simpler to just treat as ASCII if all else fails or map manually.
        // Actually, we can use String(encodings: ...)
        return String(data: data, encoding: .windowsCP1252) // Close enough often
    }
    
    private func detectDelimiter(_ line: String) -> String {
        let semi = line.filter { $0 == ";" }.count
        let comma = line.filter { $0 == "," }.count
        let tab = line.filter { $0 == "\t" }.count
        
        if semi >= comma && semi >= tab { return ";" }
        if tab > comma { return "\t" }
        return ","
    }
}

extension String {
    func toDouble() -> Double? {
        return Double(self)
    }
}
